core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java

```diff
@@ -194,8 +194,16 @@ void failTask(Task task, PayloadType payloadType, String errorMsg) {
         throw new TerminateWorkflowException(errorMsg, Workflow.WorkflowStatus.FAILED, task);
     }
 
-    private void failWorkflow(String errorMsg) {
+    @VisibleForTesting
+    void failWorkflow(Workflow workflow, PayloadType payloadType, String errorMsg) {
         LOGGER.error(errorMsg);
-        throw new TerminateWorkflowException(errorMsg);
+        workflow.setReasonForIncompletion(errorMsg);
```

Comment:
Lines L200- L206 does not look necessary as the catch block in [WorkflowExecutor.decide](https://github.com/Netflix/conductor/blob/main/core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java#L1162-L1165) and [WorkflowExecutor.terminateWorkflow](https://github.com/Netflix/conductor/blob/main/core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java#L849) set the status and reasonForIncompletion

Reply:
I was going to remove that. But then on a second thought, I think if we get any exception from uploading outputdata to external storage during the terminate step, the final status should be `failed` instead of `terminated` with specific error message reflects that. What do you think?

Reply:
I think the final status should be `TERMINATED`. A workflow is marked as FAILED, because of a Task failure or it timing out, not because of internal errors during the workflow evaluation. The `reasonForCompletion` should also be the original error that triggered the termination, not the internal error.

Reply:
Sure ya updated